File: python/729-my-calendar-i.py
class MyCalendar:

    def __init__(self):
        self.root = None

    def book(self, start: int, end: int) -> bool:
        interval = Interval(start, end)
        if self.root is None:
            self.root = interval
            return True
        return self.root.insert(interval)
        
class Interval:
    def __init__(self, start, end):
        self.start = start
        self.end = end
        self.left = None
        self.right = None
    def insert(self, interval):
        if self.end <= interval.start:
            if self.right is None:
                self.right = interval
                return True
            return self.right.insert(interval)
        elif self.start >= interval.end:
            if self.left is None:
                self.left = interval
                return True
            return self.left.insert(interval)
        else:
            return False

# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(start,end)

File: java/729-my-calendar-i.java
class MyCalendar {
    TreeMap<Integer, Integer> calendar;
    public MyCalendar() {
        calendar = new TreeMap<>();
    }
    
    public boolean book(int start, int end) {
        Integer prev = calendar.floorKey(start);
        Integer next = calendar.ceilingKey(start);
        if (prev != null && calendar.get(prev) > start) {
            return false;
        }
        if (next != null && end > next) {
            return false;
        }
        calendar.put(start, end);
        return true;
    }
    
}

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar obj = new MyCalendar();
 * boolean param_1 = obj.book(start,end);
 */